crates/starknet-devnet/tests/test_account_impersonation.rs

```diff
@@ -0,0 +1,304 @@
+            (
+                Some(ImpersonationAction::ImpersonateAccount(account.address())),
+                false,
+                TestCaseResult::Success,
+            ),
+            (None, false, TestCaseResult::Success),
+            (None, true, TestCaseResult::Failure { msg: "invalid signature".to_string() }),
+            (Some(ImpersonationAction::AutoImpersonate), true, TestCaseResult::Success),
+            (Some(ImpersonationAction::AutoImpersonate), false, TestCaseResult::Success),
+        ];
+
+        for (impersonation_action_option, do_validate, expected_result) in steps {
+            if let Some(impersonation_action) = impersonation_action_option {
+                forked_devnet.execute_impersonation_action(&impersonation_action).await.unwrap();
+            }
+
+            let simulation_result =
+                account.execute(invoke_calls.clone()).simulate(!do_validate, false).await;
+            match expected_result {
+                TestCaseResult::Success => {
+                    simulation_result.expect("Expected simulation to succeed");
+                }
+                TestCaseResult::Failure { msg } => {
+                    let err = simulation_result.err().unwrap();
+                    assert_contains(&format!("{:?}", err).to_lowercase(), &msg);
+                }
+            }
+
+            forked_devnet.restart().await.unwrap();
+        }
+    }
+
+    async fn test_declare_transaction(
+        origin_devnet: &BackgroundDevnet,
+        impersonation_actions: &[ImpersonationAction],
+        expected_result: TestCaseResult,
+    ) {
+        let forked_devnet = origin_devnet.fork().await.unwrap();
+
+        let mut account =
+            get_account_interacting_with_forked_devnet(origin_devnet, &forked_devnet).await;
+
+        for action in impersonation_actions.iter() {
+            forked_devnet.execute_impersonation_action(action).await.unwrap();
+        }
+
+        let (flattened_class, compiled_class_hash) =
+            get_simple_contract_in_sierra_and_compiled_class_hash();
+
+        account.set_block_id(BlockId::Tag(BlockTag::Latest));
+
+        let declare_result =
+            account.declare(Arc::new(flattened_class), compiled_class_hash).send().await;
+
+        match expected_result {
+            TestCaseResult::Success => {
+                crate::common::utils::assert_tx_successful(
+                    &declare_result.unwrap().transaction_hash,
+                    &forked_devnet.json_rpc_client,
+                )
+                .await;
+            }
+            TestCaseResult::Failure { msg } => {
+                let err = declare_result.err().unwrap();
+                assert_contains(&format!("{:?}", err).to_lowercase(), &msg);
+            }
+        }
+    }
+
+    async fn test_invoke_transaction(
+        origin_devnet: &BackgroundDevnet,
+        impersonation_actions: &[ImpersonationAction],
+        expected_result: TestCaseResult,
+    ) {
```

Comment:
Instead of defining a new `TestCaseResult`, wouldn't it be more in the spirit of Rust to frame the functions using it to work like this:
```rust
async fn test_invoke_transaction(
    origin_devnet: &BackgroundDevnet,
    impersonation_actions: &[ImpersonationAction]
) -> Result<(), SomeErrorType>
```
and in the calling code, if expecting success, do:
```rust
test_invoke_transaction(...).await.unwrap();
```

and if expecting failure, do:
```rust
// assert_failure could be defined to unwrap and call assert_contains
assert_failure(
    test_invoke_transaction(...).await,
    "invalid_signature"
)
```

If you feel such intervention is out of scope of this PR, it can be dealt with in another PR.

Reply:
So if the signature is invalid, validation will fail and execution result assertion won't be reached, right?

Reply:
yes